File: web/e2e/suites/tasks.mjs
/**
 * Phase 7.2 browser E2E - the task and agent session API.
 *
 * # What this suite is, and what it deliberately is not
 *
 * There is no Task UI in this phase, so nothing here clicks a button. The claim
 * under test is the API itself: that a browser, from the server's own origin,
 * can create a task, read it back, start an attempt at it, read that back, and
 * that none of it leaks across projects.
 *
 * The requests go through the page's own `fetch` rather than Node's, which is
 * the whole reason a browser suite exists for an API with no interface. A request
 * made from Node would prove the route answers; a request made from the page
 * proves the browser can reach it - same origin, no preflight the server does not
 * answer, no header the page is not allowed to set. Those are the failures a
 * person would meet and a Node-level test would not.
 *
 * # What it adds to the Go tests
 *
 * Those already pin the model, the lifecycle, the conditional update and every
 * route. What this adds is that the whole thing works over a real socket, through
 * the real server binary, against a database the same code path migrated from
 * nothing.
 *
 * # The fixture projects
 *
 * The runner registered them and started some of their runtimes before this
 * suite ran, and this suite does not start, stop or destroy anything: creating a
 * task is a record and must not touch a terminal, which is one of the claims
 * below.
 */
import { chromium } from 'playwright'

import { openPage, project, Report } from '../lib/harness.mjs'

const report = new Report('tasks')

/**
 * call performs an API request from inside the page.
 *
 * It returns the status and the parsed body rather than throwing, because most
 * of what this suite checks is a refusal: a helper that threw on a 409 would make
 * "the transition was refused" read as an error in the test.
 */
async function call(page, method, path, body) {
  return page.evaluate(
    async ([method, path, body]) => {
      const init = { method }
      if (body !== null) {
        init.body = JSON.stringify(body)
        init.headers = { 'Content-Type': 'application/json' }
      }
      const response = await fetch(`/api${path}`, init)
      const text = await response.text()
      let parsed = null
      try {
        parsed = text === '' ? null : JSON.parse(text)
      } catch {
        // A refusal the server did not write - a 405 from the router, say - is
        // still an answer, and the status is what the check reads.
        parsed = { raw: text }
      }
      return { status: response.status, body: parsed }
    },
    [method, path, body ?? null],
  )
}

/** errorCode reads the stable code out of a failing response. */
function errorCode(response) {
  return response.body?.error?.code ?? ''
}

/** runtimeIdOf is the runtime a project's session runs in. */
const runtimeIdOf = (entry) => `amx-${entry.id}`

async function main() {
  const browser = await chromium.launch({ channel: 'chrome', headless: true })
  try {
    // The two projects the claims are made against.
    //
    // `a` is the flagship: a real Claude session, alive while this runs, so "a
    // task is not a runtime" is asked of a runtime that genuinely exists rather
    // than of one that was never started. `b` is an ordinary shell, which makes
    // the isolation claim cross the two kinds of fixture instead of comparing a
    // project with itself.
    const a = project('claude')
    const b = project('shell')

    const only = process.env.AMX_E2E_ONLY ?? ''
    const sections = [
      ['create', creatingATask],
      ['read', readingATaskBack],
      ['lifecycle', theLifecycle],
      ['session', anAttemptAtATask],
      ['isolation', projectIsolation],
      ['separation', aTaskIsNotARuntime],
    ]

    for (const [name, section] of sections) {
      if (only !== '' && !name.includes(only)) continue
      await section(browser, a, b)
    }
  } finally {
    await browser.close()
  }

  process.exit(report.summary() ? 0 : 1)
}

/** Creating a task: what comes back, and what does not. */
async function creatingATask(browser, a) {
  console.log('\n--- creating a task ---')
  const { context, page } = await openPage(browser)

  const created = await call(page, 'POST', `/projects/${a.id}/tasks`, {
    title: 'Implement Controller Viewer',
  })
  report.check(
    'a browser can create a task through the server it was served from',
    created.status === 201 && created.body?.task?.id?.startsWith('task_'),
    `status ${created.status}, id ${created.body?.task?.id}`,
  )

  const task = created.body?.task ?? {}
  report.check(
    'the task belongs to the project in the path and starts in CREATED',
    task.projectId === a.id && task.status === 'CREATED',
    `projectId ${task.projectId}, status ${task.status}`,
  )
  report.check(
    'the server produced the timestamps, in UTC',
    typeof task.createdAt === 'string' && task.createdAt.endsWith('Z') && task.completedAt === null,
    `createdAt ${task.createdAt}, completedAt ${task.completedAt}`,
  )

  // The title is what the person typed and it comes back unchanged. This is not
  // a security test - the Go tests cover that - it is that the round trip through
  // a real HTTP stack preserves text rather than mangling it.
  const typed = '实现控制器查看器 🚀 "quoted"'
  const odd = await call(page, 'POST', `/projects/${a.id}/tasks`, { title: typed })
  report.check(
    'a title with non-ASCII text and quotes round-trips unchanged',
    odd.status === 201 && odd.body?.task?.title === typed,
    `got ${JSON.stringify(odd.body?.task?.title)}`,
  )

  const tooLong = await call(page, 'POST', `/projects/${a.id}/tasks`, { title: 'x'.repeat(201) })
  report.check(
    'an over-long title is refused rather than stored',
    tooLong.status === 400 && errorCode(tooLong) === 'invalid_task_title',
    `status ${tooLong.status}, code ${errorCode(tooLong)}`,
  )

  const unknownProject = await call(page, 'POST', '/projects/p_ffffffffffffffffffff/tasks', {
    title: 'work',
  })
  report.check(
    'a task cannot be created under a project that does not exist',
    unknownProject.status === 404 && errorCode(unknownProject) === 'project_not_found',
    `status ${unknownProject.status}, code ${errorCode(unknownProject)}`,
  )

  await context.close()
}

/** Reading a task back, by id and in its project's listing. */
async function readingATaskBack(browser, a) {
  console.log('\n--- reading a task back ---')
  const { context, page } = await openPage(browser)

  const title = `read back ${Date.now()}`
  const created = await call(page, 'POST', `/projects/${a.id}/tasks`, { title })
  const id = created.body?.task?.id ?? ''

  const fetched = await call(page, 'GET', `/tasks/${id}`)
  report.check(
    'a task is readable by its own id',
    fetched.status === 200 && fetched.body?.task?.id === id && fetched.body?.task?.title === title,
    `status ${fetched.status}, title ${JSON.stringify(fetched.body?.task?.title)}`,
  )

  const listed = await call(page, 'GET', `/projects/${a.id}/tasks?limit=500`)
  const found = (listed.body?.tasks ?? []).some((task) => task.id === id)
  report.check(
    'the task appears in its project’s listing',
    listed.status === 200 && found,
    `status ${listed.status}, ${listed.body?.count} task(s) in the listing`,
  )

  const filtered = await call(page, 'GET', `/projects/${a.id}/tasks?status=NOPE`)
  report.check(
    'an unknown status filter is refused rather than answered with an empty list',
    filtered.status === 400 && errorCode(filtered) === 'invalid_input',
    `status ${filtered.status}, code ${errorCode(filtered)}`,
  )

  const missing = await call(page, 'GET', '/tasks/task_ffffffffffffffffffffffff')
  report.check(
    'an unknown task is a 404 rather than an empty task',
    missing.status === 404 && errorCode(missing) === 'task_not_found',
    `status ${missing.status}, code ${errorCode(missing)}`,
  )

  await context.close()
}

/** The lifecycle, over HTTP: what moves, and what cannot. */
async function theLifecycle(browser, a) {
  console.log('\n--- the lifecycle ---')
  const { context, page } = await openPage(browser)

  const created = await call(page, 'POST', `/projects/${a.id}/tasks`, { title: 'lifecycle' })
  const id = created.body?.task?.id ?? ''

  const running = await call(page, 'PATCH', `/tasks/${id}`, { status: 'RUNNING' })
  report.check(
    'a task moves from CREATED to RUNNING',
    running.status === 200 && running.body?.task?.status === 'RUNNING',
    `status ${running.status}, task status ${running.body?.task?.status}`,
  )

  const completed = await call(page, 'PATCH', `/tasks/${id}`, { status: 'COMPLETED' })
  report.check(
    'completing a task records when it completed',
    completed.status === 200 && completed.body?.task?.completedAt !== null,
    `status ${completed.status}, completedAt ${completed.body?.task?.completedAt}`,
  )

  // The one edge the directive names by hand. A task reported as finished and
  // then reported as running again would make the first report false.
  const backwards = await call(page, 'PATCH', `/tasks/${id}`, { status: 'RUNNING' })
  report.check(
    'COMPLETED is final: moving back to RUNNING is refused',
    backwards.status === 409 && errorCode(backwards) === 'invalid_status_transition',
    `status ${backwards.status}, code ${errorCode(backwards)}`,
  )

  const after = await call(page, 'GET', `/tasks/${id}`)
  report.check(
    'the refused transition changed nothing',
    after.body?.task?.status === 'COMPLETED',
    `status is now ${after.body?.task?.status}`,
  )

  const renamed = await call(page, 'PATCH', `/tasks/${id}`, { title: 'lifecycle, renamed' })
  report.check(
    'a task can be renamed without changing its status',
    renamed.status === 200 &&
      renamed.body?.task?.title === 'lifecycle, renamed' &&
      renamed.body?.task?.status === 'COMPLETED',
    `status ${renamed.status}, title ${JSON.stringify(renamed.body?.task?.title)}`,
  )

  const noDelete = await call(page, 'DELETE', `/tasks/${id}`)
  report.check(
    'there is no delete for a task',
    noDelete.status === 404 || noDelete.status === 405,
    `DELETE returned ${noDelete.status}`,
  )

  await context.close()
}

/** An attempt at a task: created with no runtime, then bound to one. */
async function anAttemptAtATask(browser, a, b) {
  console.log('\n--- an attempt at a task ---')
  const { context, page } = await openPage(browser)

  const created = await call(page, 'POST', `/projects/${a.id}/tasks`, { title: 'attempts' })
  const taskId = created.body?.task?.id ?? ''

  const first = await call(page, 'POST', `/tasks/${taskId}/sessions`)
  const session = first.body?.session ?? {}
  report.check(
    'an attempt is created with no runtime and no start time',
    first.status === 201 &&
      session.id?.startsWith('sess_') &&
      session.runtimeId === undefined &&
      session.startedAt === null,
    `status ${first.status}, runtimeId ${JSON.stringify(session.runtimeId)}, ` +
      `startedAt ${session.startedAt}`,
  )

  const running = await call(page, 'PATCH', `/sessions/${session.id}`, { status: 'RUNNING' })
  report.check(
    'starting an attempt records when it started',
    running.status === 200 &&
      running.body?.session?.status === 'RUNNING' &&
      running.body?.session?.startedAt !== null,
    `status ${running.status}, startedAt ${running.body?.session?.startedAt}`,
  )

  // The runtime is attached afterwards, and it is deliberately not required to
  // exist: a runtime is destroyed and forgotten while the record of the attempt
  // that used it remains.
  const runtimeId = runtimeIdOf(a)
  const attached = await call(page, 'PATCH', `/sessions/${session.id}`, { runtimeId })
  report.check(
    'a runtime is bound to the attempt after the fact',
    attached.status === 200 && attached.body?.session?.runtimeId === runtimeId,
    `status ${attached.status}, runtimeId ${attached.body?.session?.runtimeId}`,
  )

  const rebound = await call(page, 'PATCH', `/sessions/${session.id}`, {
    runtimeId: runtimeIdOf(b),
  })
  report.check(
    'a bound runtime cannot be replaced',
    rebound.status === 409 && errorCode(rebound) === 'status_conflict',
    `status ${rebound.status}, code ${errorCode(rebound)}`,
  )

  const ended = await call(page, 'PATCH', `/sessions/${session.id}`, { status: 'COMPLETED' })
  report.check(
    'completing an attempt records when it ended and keeps its start',
    ended.status === 200 &&
      ended.body?.session?.endedAt !== null &&
      ended.body?.session?.startedAt !== null,
    `status ${ended.status}, endedAt ${ended.body?.session?.endedAt}`,
  )

  // A second attempt is a second session. Neither replaces the other, which is
  // what this level exists for.
  const second = await call(page, 'POST', `/tasks/${taskId}/sessions`)
  const listed = await call(page, 'GET', `/tasks/${taskId}/sessions`)
  report.check(
    'a second attempt is a second attempt, and both are kept',
    second.status === 201 && listed.body?.count === 2,
    `${listed.body?.count} attempt(s) after the second one`,
  )

  const waiting = await call(page, 'PATCH', `/sessions/${second.body?.session?.id}`, {
    status: 'WAITING',
  })
  report.check(
    'WAITING is a task status and not an attempt status',
    waiting.status === 400 && errorCode(waiting) === 'invalid_input',
    `status ${waiting.status}, code ${errorCode(waiting)}`,
  )

  const noDelete = await call(page, 'DELETE', `/sessions/${session.id}`)
  report.check(
    'there is no delete for an attempt',
    noDelete.status === 404 || noDelete.status === 405,
    `DELETE returned ${noDelete.status}`,
  )

  await context.close()
}

/** Nothing leaks between projects. */
async function projectIsolation(browser, a, b) {
  console.log('\n--- project isolation ---')
  const { context, page } = await openPage(browser)

  const mine = await call(page, 'POST', `/projects/${a.id}/tasks`, { title: `A only ${Date.now()}` })
  const theirs = await call(page, 'POST', `/projects/${b.id}/tasks`, {
    title: `B only ${Date.now()}`,
  })

  const listedForB = await call(page, 'GET', `/projects/${b.id}/tasks?limit=500`)
  const leaked = (listedForB.body?.tasks ?? []).some(
    (task) => task.id === mine.body?.task?.id || task.projectId === a.id,
  )
  report.check(
    'one project’s listing does not contain another’s tasks',
    listedForB.status === 200 && !leaked,
    `${listedForB.body?.count} task(s) in B's listing, leaked: ${leaked}`,
  )

  // A task is reached by its own id, so it is readable wherever the id is known
  // - that is what a url is. What must hold is that it still says which project
  // it belongs to, wherever it is read from.
  const fetched = await call(page, 'GET', `/tasks/${mine.body?.task?.id}`)
  report.check(
    'a task carries its project with it wherever it is read',
    fetched.status === 200 && fetched.body?.task?.projectId === a.id,
    `projectId ${fetched.body?.task?.projectId}, want ${a.id}`,
  )

  report.check(
    'the two projects’ tasks are different tasks',
    mine.body?.task?.id !== theirs.body?.task?.id,
    `${mine.body?.task?.id} vs ${theirs.body?.task?.id}`,
  )

  await context.close()
}

/**
 * A task is not a runtime.
 *
 * §二十九 says creating a task must not start anything, and §五 is why: a task is
 * the goal, a runtime is one way of working on it, and a task that needed a
 * process to exist would not be a goal at all.
 */
async function aTaskIsNotARuntime(browser, a) {
  console.log('\n--- a task is not a runtime ---')
  const { context, page } = await openPage(browser)

  const before = await call(page, 'GET', `/projects/${a.id}/runtime`)
  report.check(
    'the fixture project has a runtime to ask about',
    before.status === 200 && typeof before.body?.runtime?.state === 'string',
    `status ${before.status}, state ${before.body?.runtime?.state}`,
  )

  const created = await call(page, 'POST', `/projects/${a.id}/tasks`, { title: 'starts nothing' })
  const taskId = created.body?.task?.id ?? ''

  const after = await call(page, 'GET', `/projects/${a.id}/runtime`)
  report.check(
    'creating a task does not start, stop or destroy a runtime',
    before.body?.runtime?.state === after.body?.runtime?.state &&
      before.body?.runtime?.sessionAlive === after.body?.runtime?.sessionAlive,
    `state ${before.body?.runtime?.state} -> ${after.body?.runtime?.state}, ` +
      `alive ${before.body?.runtime?.sessionAlive} -> ${after.body?.runtime?.sessionAlive}`,
  )

  const sessions = await call(page, 'GET', `/tasks/${taskId}/sessions`)
  report.check(
    'a new task has no attempts until one is asked for',
    sessions.status === 200 && sessions.body?.count === 0,
    `${sessions.body?.count} attempt(s)`,
  )

  // The tasks recorded by this suite are still there and still reachable, which
  // is the other half of "a task outlives the process": whatever happens to a
  // runtime, the record of the work is not what goes with it.
  const listed = await call(page, 'GET', `/projects/${a.id}/tasks?limit=500`)
  report.check(
    'the project’s tasks survive whatever happens to its runtime',
    listed.status === 200 && (listed.body?.count ?? 0) > 0,
    `${listed.body?.count} task(s) recorded`,
  )

  await context.close()
}

await main()

File: internal/httpapi/tasks.go
package httpapi

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/kutonlagos/agentmux/internal/task"
)

// This file is the task and agent session API.
//
// What it is not is as much of the design as what it is:
//
//   - There is no DELETE on either resource. Nothing in this build removes a
//     task or an attempt, and a task that was cancelled is a record of work
//     somebody decided not to do - which is worth keeping. The foreign keys in
//     0004 and 0005 decide what would happen the day something does delete;
//     this phase does not open that door.
//   - There is no endpoint that starts anything. Creating a task records that
//     somebody wants work done. It starts no runtime, launches no agent and
//     writes no prompt: the two halves of AgentMux meet at the runtime API,
//     which already exists, and joining them is a later phase.
//   - A session is created with no runtime, and one is attached afterwards.
//     That window is a state the model names rather than hides, and POST here
//     deliberately does not accept a runtimeId - a caller that wants both makes
//     two requests, and the first one is true on its own. See
//     docs/TASK_MODEL.md §4.
//
// The project half of the path is the project the work belongs to, and the
// project is resolved by the service before anything is written, so a task
// cannot be created under a project that does not exist. A request body that
// tries to name one is refused by the unknown-field check rather than quietly
// ignored - see createTaskRequest.

// tasksTimeout bounds a task or session request.
//
// These are metadata operations over a local SQLite file, not runtime ones, so
// they get the metadata budget rather than the longer one a process launch
// needs.
const tasksTimeout = 10 * time.Second

// taskResponse is the body of the endpoints that return a single task.
type taskResponse struct {
	Task *task.Task `json:"task"`
}

// taskListResponse is the body of GET /api/projects/{id}/tasks.
type taskListResponse struct {
	Tasks []*task.Task `json:"tasks"`
	Count int          `json:"count"`
}

// sessionResponse is the body of the endpoints that return a single session.
type sessionResponse struct {
	Session *task.AgentSession `json:"session"`
}

// sessionListResponse is the body of GET /api/tasks/{id}/sessions.
type sessionListResponse struct {
	Sessions []*task.AgentSession `json:"sessions"`
	Count    int                  `json:"count"`
}

// createTaskRequest is the body of POST /api/projects/{id}/tasks.
//
// It has one field, and the project is not one of them. The project is in the
// path, and a body that also names one is an unknown field and refused. That is
// the point: two places to say which project a task belongs to is one place too
// many, and the disagreement between them would have to be resolved by a rule
// nobody wrote down.
type createTaskRequest struct {
	Title string `json:"title"`
}

// handleCreateTask implements POST /api/projects/{id}/tasks.
//
// The task begins in CREATED, with no session. Nothing is started.
func (s *Server) handleCreateTask(w http.ResponseWriter, r *http.Request) {
	if !s.requireTasks(w) {
		return
	}
	ctx, cancel := s.contextWithTimeout(r, tasksTimeout)
	defer cancel()

	var req createTaskRequest
	if err := decodeJSON(w, r, &req); err != nil {
		writeError(w, http.StatusBadRequest, CodeInvalidRequest, err.Error(), nil)
		return
	}

	t, err := s.tasks.CreateTask(ctx, task.CreateTaskInput{
		ProjectID: r.PathValue("id"),
		Title:     req.Title,
	})
	if err != nil {
		writeServiceError(w, s.log, err)
		return
	}
	writeJSON(w, s.log, http.StatusCreated, taskResponse{Task: t})
}

// handleListTasks implements GET /api/projects/{id}/tasks.
//
// The project is resolved before the list is read, so an unknown id is answered
// as a missing project rather than as a project with no tasks. Those two answers
// mean different things, and an empty list is the one a client cannot tell apart
// from a project that exists and has nothing in it.
func (s *Server) handleListTasks(w http.ResponseWriter, r *http.Request) {
	if !s.requireTasks(w) {
		return
	}
	ctx, cancel := s.contextWithTimeout(r, tasksTimeout)
	defer cancel()

	limit, ok := limitFrom(w, r)
	if !ok {
		return
	}

	tasks, err := s.tasks.ListTasks(ctx, task.ListTasksInput{
		ProjectID: r.PathValue("id"),
		// Passed through untrimmed-but-trimmed and unmodified: a status the model
		// does not know is refused with the list it does know, which is a more
		// useful answer than a silently empty page.
		Status: strings.TrimSpace(r.URL.Query().Get("status")),
		Limit:  limit,
	})
	if err != nil {
		writeServiceError(w, s.log, err)
		return
	}
	writeJSON(w, s.log, http.StatusOK, taskListResponse{Tasks: tasks, Count: len(tasks)})
}

// handleGetTask implements GET /api/tasks/{id}.
func (s *Server) handleGetTask(w http.ResponseWriter, r *http.Request) {
	if !s.requireTasks(w) {
		return
	}
	ctx, cancel := s.contextWithTimeout(r, tasksTimeout)
	defer cancel()

	t, err := s.tasks.GetTask(ctx, r.PathValue("id"))
	if err != nil {
		writeServiceError(w, s.log, err)
		return
	}
	writeJSON(w, s.log, http.StatusOK, taskResponse{Task: t})
}

// updateTaskRequest is the body of PATCH /api/tasks/{id}.
//
// Both fields are optional and both are pointers into a Present flag, because
// three things have to be distinguishable here: a field that was not sent, a
// field sent as a string, and a field sent as null. A plain *string collapses
// the first and the third, and the third is a mistake worth reporting rather
// than treating as "leave it alone".
type updateTaskRequest struct {
	Title  optionalString `json:"title"`
	Status optionalString `json:"status"`
}

// handleUpdateTask implements PATCH /api/tasks/{id}.
//
// The status is changed through this endpoint rather than through a POST to a
// /status sub-resource, because a status is a property of a task and not a
// resource of its own. A PATCH also makes the rule visible in the request: the
// caller says what the task should become, not what should happen to it.
//
// It is the only route that changes a task's status, and it cannot bypass the
// lifecycle: the service checks the transition against the status it reads and
// writes conditionally on that status still holding. A COMPLETED task cannot be
// moved back to RUNNING by any request this server accepts.
func (s *Server) handleUpdateTask(w http.ResponseWriter, r *http.Request) {
	if !s.requireTasks(w) {
		return
	}
	ctx, cancel := s.contextWithTimeout(r, tasksTimeout)
	defer cancel()

	var req updateTaskRequest
	if err := decodeJSON(w, r, &req); err != nil {
		writeError(w, http.StatusBadRequest, CodeInvalidRequest, err.Error(), nil)
		return
	}

	in := task.UpdateTaskInput{}
	if req.Title.Present {
		if req.Title.Value == nil {
			writeError(w, http.StatusBadRequest, CodeInvalidRequest,
				"title must be a string; a task title cannot be cleared", nil)
			return
		}
		in.Title = req.Title.Value
	}
	if req.Status.Present {
		if req.Status.Value == nil {
			writeError(w, http.StatusBadRequest, CodeInvalidRequest,
				fmt.Sprintf("status must be a string; a task status is one of %s",
					strings.Join(task.TaskStatuses(), ", ")), nil)
			return
		}
		in.Status = req.Status.Value
	}
	if in.Title == nil && in.Status == nil {
		writeError(w, http.StatusBadRequest, CodeInvalidRequest,
			"the request must set title, status, or both", nil)
		return
	}

	t, err := s.tasks.UpdateTask(ctx, r.PathValue("id"), in)
	if err != nil {
		writeServiceError(w, s.log, err)
		return
	}
	writeJSON(w, s.log, http.StatusOK, taskResponse{Task: t})
}

// createSessionRequest is the body of POST /api/tasks/{id}/sessions, and it has
// no fields.
//
// The task is in the path and the runtime is attached by a later request, so an
// attempt needs nothing else to exist. The body is still decoded rather than
// ignored, so that a client which sends a field this API does not accept is
// told so - and the body is optional, because "start an attempt at this task"
// is a complete request with nothing in it.
type createSessionRequest struct{}

// handleCreateSession implements POST /api/tasks/{id}/sessions.
//
// The session begins in CREATED with no runtime. That is not an oversight and it
// is not a gap to be filled by a transaction: creating an attempt and starting a
// process are two acts, and the second one launches something that can fail on
// its own. A single transaction across both would hold a write lock open for the
// length of a process launch and would make a tmux that refused to start undo
// the record that somebody wanted this work done at all.
func (s *Server) handleCreateSession(w http.ResponseWriter, r *http.Request) {
	if !s.requireTasks(w) {
		return
	}
	ctx, cancel := s.contextWithTimeout(r, tasksTimeout)
	defer cancel()

	var req createSessionRequest
	if err := decodeJSON(w, r, &req); err != nil && !errors.Is(err, errEmptyBody) {
		writeError(w, http.StatusBadRequest, CodeInvalidRequest, err.Error(), nil)
		return
	}

	sess, err := s.tasks.CreateSession(ctx, task.CreateSessionInput{
		TaskID: r.PathValue("id"),
	})
	if err != nil {
		writeServiceError(w, s.log, err)
		return
	}
	writeJSON(w, s.log, http.StatusCreated, sessionResponse{Session: sess})
}

// handleListSessions implements GET /api/tasks/{id}/sessions.
//
// One task's attempts, newest first. The task is resolved first, so an unknown
// id is a 404 about the task rather than an empty list of attempts.
func (s *Server) handleListSessions(w http.ResponseWriter, r *http.Request) {
	if !s.requireTasks(w) {
		return
	}
	ctx, cancel := s.contextWithTimeout(r, tasksTimeout)
	defer cancel()

	limit, ok := limitFrom(w, r)
	if !ok {
		return
	}

	sessions, err := s.tasks.ListSessions(ctx, task.ListSessionsInput{
		TaskID: r.PathValue("id"),
		Limit:  limit,
	})
	if err != nil {
		writeServiceError(w, s.log, err)
		return
	}
	writeJSON(w, s.log, http.StatusOK,
		sessionListResponse{Sessions: sessions, Count: len(sessions)})
}

// handleGetSession implements GET /api/sessions/{id}.
func (s *Server) handleGetSession(w http.ResponseWriter, r *http.Request) {
	if !s.requireTasks(w) {
		return
	}
	ctx, cancel := s.contextWithTimeout(r, tasksTimeout)
	defer cancel()

	sess, err := s.tasks.GetSession(ctx, r.PathValue("id"))
	if err != nil {
		writeServiceError(w, s.log, err)
		return
	}
	writeJSON(w, s.log, http.StatusOK, sessionResponse{Session: sess})
}

// updateSessionRequest is the body of PATCH /api/sessions/{id}.
type updateSessionRequest struct {
	Status optionalString `json:"status"`

	// RuntimeID binds the attempt to the runtime it ran in.
	//
	// It is set here rather than at creation because the two are different
	// facts recorded at different times: the attempt exists first, and the
	// runtime it used is known when one is started. It may be set once - see
	// the repository's AttachSessionRuntime - and it is deliberately not
	// checked for existence: a runtime can be destroyed while the attempt that
	// used it remains, which is the whole reason it is not a foreign key.
	RuntimeID optionalString `json:"runtimeId"`
}

// handleUpdateSession implements PATCH /api/sessions/{id}.
//
// Two fields, and the order they are applied in is a decision.
//
// The status is written first, because it is the write that can lose a race: the
// service validates the transition against the status it read and the update is
// conditional on that status still holding. A caller told it lost a race is told
// so before a runtime has been attached on its behalf, which keeps the refusal
// honest - nothing about the request was applied.
//
// The reverse order would be worse in the case that matters: a runtime attached
// to an attempt whose status change then failed would leave a link recording
// that this attempt ran in a runtime, established by a request that was refused.
func (s *Server) handleUpdateSession(w http.ResponseWriter, r *http.Request) {
	if !s.requireTasks(w) {
		return
	}
	ctx, cancel := s.contextWithTimeout(r, tasksTimeout)
	defer cancel()

	var req updateSessionRequest
	if err := decodeJSON(w, r, &req); err != nil {
		writeError(w, http.StatusBadRequest, CodeInvalidRequest, err.Error(), nil)
		return
	}

	id := r.PathValue("id")
	sess, err := s.tasks.GetSession(ctx, id)
	if err != nil {
		writeServiceError(w, s.log, err)
		return
	}

	switch {
	case req.Status.Present && req.Status.Value == nil:
		writeError(w, http.StatusBadRequest, CodeInvalidRequest,
			fmt.Sprintf("status must be a string; an agent session status is one of %s",
				strings.Join(task.AgentSessionStatuses(), ", ")), nil)
		return
	case req.RuntimeID.Present && req.RuntimeID.Value == nil:
		writeError(w, http.StatusBadRequest, CodeInvalidRequest,
			"runtimeId must be a string; a runtime is attached, never detached", nil)
		return
	case req.RuntimeID.Present && strings.TrimSpace(*req.RuntimeID.Value) == "":
		// A blank runtime id is refused here rather than by the service, because
		// the no-op comparison below would otherwise swallow it: a session that
		// has no runtime would be told it now has the one it asked for, and
		// nothing would have been recorded. There is no runtime whose id is the
		// empty string, so this is never a request anybody made on purpose.
		writeError(w, http.StatusBadRequest, CodeInvalidRequest,
			"runtimeId must name a runtime; a runtime id is its session name, "+
				"which is \"amx-\" followed by a project id", nil)
		return
	case !req.Status.Present && !req.RuntimeID.Present:
		writeError(w, http.StatusBadRequest, CodeInvalidRequest,
			"the request must set status, runtimeId, or both", nil)
		return
	}

	if req.Status.Present && *req.Status.Value != sess.Status {
		if sess, err = s.tasks.UpdateSessionStatus(ctx, id, *req.Status.Value); err != nil {
			writeServiceError(w, s.log, err)
			return
		}
	}
	if req.RuntimeID.Present && *req.RuntimeID.Value != sess.RuntimeID {
		if sess, err = s.tasks.AttachSessionRuntime(ctx, id, *req.RuntimeID.Value); err != nil {
			writeServiceError(w, s.log, err)
			return
		}
	}
	writeJSON(w, s.log, http.StatusOK, sessionResponse{Session: sess})
}

// optionalString is a string field that can tell "not sent" from "sent as
// null".
//
// encoding/json decodes both an absent field and an explicit null into a nil
// pointer, and the two are different here: absent means leave the field alone,
// while null means the caller asked to clear something that cannot be cleared.
// Reporting the second as the first would silently accept a request to unset a
// task's title.
type optionalString struct {
	// Present is true when the field appeared in the body at all.
	Present bool

	// Value is the string that was sent, or nil for an explicit null.
	Value *string
}

// UnmarshalJSON records that the field was sent, then decodes it.
func (v *optionalString) UnmarshalJSON(data []byte) error {
	v.Present = true
	if strings.TrimSpace(string(data)) == "null" {
		v.Value = nil
		return nil
	}
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}
	v.Value = &s
	return nil
}

// limitFrom reads the `limit` query parameter, answering the request itself when
// it is malformed.
//
// The second result reports whether the caller should carry on. Zero means "the
// service's default": the parameter was absent, and a default is the service's
// to choose rather than this layer's to restate.
func limitFrom(w http.ResponseWriter, r *http.Request) (int, bool) {
	raw := strings.TrimSpace(r.URL.Query().Get("limit"))
	if raw == "" {
		return 0, true
	}
	value, err := strconv.Atoi(raw)
	if err != nil {
		writeError(w, http.StatusBadRequest, CodeInvalidRequest,
			fmt.Sprintf("limit must be a whole number, not %q", raw), nil)
		return 0, false
	}
	if value < 1 {
		// Zero is not "no limit" and is not "the default". A caller that asked
		// for nothing has made a mistake, and answering with a page would hide
		// it.
		writeError(w, http.StatusBadRequest, CodeInvalidRequest,
			fmt.Sprintf("limit must be at least 1, not %d", value), nil)
		return 0, false
	}
	// Above the ceiling the value is passed through and the service clamps it,
	// which is where the ceiling is declared. A caller asking for a thousand
	// wants as much as it can have, not a refusal.
	return value, true
}

// requireTasks refuses a task request on a server built without a task service.
//
// The service is optional in Options for the same reason the event log and the
// terminal hub are: a test of the REST surface that is not about tasks should
// not have to build one. A server without it explains itself rather than
// panicking.
func (s *Server) requireTasks(w http.ResponseWriter) bool {
	if s.tasks == nil {
		writeError(w, http.StatusServiceUnavailable, CodeInternal,
			"this server was started without a task model", nil)
		return false
	}
	return true
}
